Source/NSPopover.m

```diff
@@ -34,101 +34,290 @@
 #import "AppKit/NSPopover.h"
 #import "AppKit/NSViewController.h"
 #import "AppKit/NSView.h"
-#import "AppKit/NSWindow.h"
+#import "AppKit/NSPanel.h"
+#import "AppKit/NSNibLoading.h"
+#import "AppKit/NSStoryboard.h"
+#import "AppKit/NSGraphics.h"
 
+// Popover private classes
+
+@interface GSPopoverView : NSView
+{
+  NSView *_contentView;
+}
+@end
+
+@implementation GSPopoverView
+
+- (instancetype) initWithFrame: (NSRect)frame
+{
+  self = [super initWithFrame: frame];
+  if (self)
+    {
+      [self setAutoresizingMask: NSViewWidthSizable | NSViewHeightSizable];
+    }
+  return self;
+}
+
+- (void) setContentView: (NSView *)view
+{
+  if (view != nil)
+    {
+      ASSIGN(_contentView, view);
+      [_contentView setAutoresizingMask: NSViewWidthSizable | NSViewHeightSizable];
+      [self addSubview: _contentView];
+    }
+  else
+    {
+      [_contentView removeFromSuperview];
+    }
+}
+
+- (void) drawRect: (NSRect)dirtyRect
+{
+  NSRectEdge sides[] = {NSMinXEdge, NSMaxYEdge, NSMaxXEdge, NSMinYEdge};
+  NSColor *black = [NSColor blackColor];
+  NSColor *white = [NSColor whiteColor];
+  NSColor *colors[] = {white, white, black, black};
+  NSRect bounds = [self bounds];
+
+  [super drawRect: dirtyRect];
+  NSDrawColorTiledRects(bounds, bounds, sides, colors, 4);
+}
+
+@end
+
+@interface GSPopoverPanel : NSPanel
+@end
+
+@implementation GSPopoverPanel
+
+- (id) initWithContentRect: (NSRect)contentRect
+                 styleMask: (NSUInteger)aStyle
+                   backing: (NSBackingStoreType)bufferingType
+                     defer: (BOOL)flag
+{
+  self = [super initWithContentRect: contentRect
+                          styleMask: aStyle
+                            backing: bufferingType
+                              defer: flag];
+  if (self)
+    {
+      [super setContentView: AUTORELEASE([[GSPopoverView alloc] initWithFrame: contentRect])];
+    }
+  return self;
+}
+
+- (void) orderOut: (id)sender
+{
+  [super orderOut: sender];
+  [self close];
+}
+
+- (BOOL) canBecomeKeyWindow 
+{
+  return NO;
+}
+
+- (BOOL) canBecomeMainWindow 
+{
+  return NO;
+}
+
+- (void) setContentView: (NSView *)view
+{
+  [[super contentView] addSubview: view];
+}
+
+- (NSView *) contentView
+{
+  NSArray *subviews = [[super contentView] subviews];
+
+  if ([subviews count] == 0)
+    {
+      return nil;
+    }
+  
+  return [subviews objectAtIndex: 0];
+}
+
+@end
 
 /* Class */
 @implementation NSPopover
 
 /* Properties */
+
+/**
```

Comment:
Normally we would add such comments to the header file.

Reply:
Comments moved to header.